drivers/gpio/gpio_pca95xx.c

```diff
@@ -448,12 +484,149 @@ static int gpio_pca95xx_port_toggle_bits(const struct device *dev,
 	return ret;
 }
 
+static void gpio_pca95xx_interrupt_worker(struct k_work *work)
+{
+	struct gpio_pca95xx_drv_data * const drv_data = CONTAINER_OF(
+		work, struct gpio_pca95xx_drv_data, interrupt_worker);
+	uint16_t input_status, input_cache;
+	uint32_t enabled_int = 0;
+	int ret;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	input_cache = drv_data->reg_cache.input;
+
+	ret = update_input_regs(drv_data->instance, &input_status);
+	if (ret == 0) {
+		/* Note: PCA Interrupt status is cleared by reading inputs */
+
+		enabled_int = ((input_cache ^ input_status) &
+			       drv_data->pin_callback_enables);
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	gpio_fire_callbacks(&drv_data->callbacks, drv_data->instance,
+			    enabled_int);
+}
+
+static void gpio_pca95xx_interrupt_callback(const struct device *dev,
+					    struct gpio_callback *cb,
+					    gpio_port_pins_t pins)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		CONTAINER_OF(cb, struct gpio_pca95xx_drv_data, gpio_callback);
+
+	/* Cannot read PCA95xx registers from ISR context, queue worker */
+	k_work_submit(&drv_data->interrupt_worker);
+}
+
 static int gpio_pca95xx_pin_interrupt_configure(const struct device *dev,
 						  gpio_pin_t pin,
 						  enum gpio_int_mode mode,
 						  enum gpio_int_trig trig)
 {
-	return -ENOTSUP;
+	const struct gpio_pca95xx_config * const config = dev->config;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->data;
+	const struct device *int_gpio_dev;
+	int ret = 0;
+	uint16_t reg;
+
+	/* Check if GPIO port supports interrupts */
+	if ((config->capabilities & PCA_HAS_INTERRUPT) == 0U) {
+		return -ENOTSUP;
+	}
+
+	/* Check for an invalid pin number */
+	if (pin > config->common.port_pin_mask) {
```

Comment:
This doesn't seem right.  Isn't `pin` an ordinal in `[0,16)` while `port_pin_mask` is a bit mask `0xFFFF`?

Reply:
indeed a bug. fixed